lib/Dialect/Calyx/Transforms/CompileControl.cpp

```diff
@@ -0,0 +1,190 @@
+  return builder.create<hw::ConstantOp>(
+      loc, APInt(bitWidth, value, /*unsigned=*/true));
+}
+
+/// Given some number of states, returns the necessary bit width
+/// TODO(Calyx): Probably a better built-in operation?
+static size_t getNecessaryBitWidth(size_t numStates) {
+  APInt apNumStates(64, numStates);
+  size_t log2 = apNumStates.ceilLogBase2();
+  return log2 > 1 ? log2 : 1;
+}
+
+/// Creates a RegisterOp, with input and output port bit widths defined by
+/// `width`.
+static RegisterOp createRegister(OpBuilder &builder, ComponentOp &component,
+                                 size_t width, StringRef name) {
+  auto *context = builder.getContext();
+  builder.setInsertionPointToStart(component.getBody());
+  return builder.create<RegisterOp>(component->getLoc(),
+                                    StringAttr::get(context, name), width);
+}
+
+/// Generates a latency-insensitive FSM to realize a sequential operation. This
+/// is done by initializing GroupGoOp values for the enabled groups in the
+/// SeqOp, and then creating a new Seq GroupOp with the given FSM. Each step in
+/// the FSM is guarded by the done operation of the group currently being
+/// executed. After the group is complete, the FSM is incremented. This SeqOp is
+/// then replaced in the control with an Enable statement referring to the new
+/// Seq GroupOp.
+static void visitSeqOp(SeqOp &seq, ComponentOp &component) {
+  auto wires = component.getWiresOp();
+  Block *wiresBody = wires.getBody();
+
+  auto &seqOps = seq.getBody()->getOperations();
+  assert(llvm::all_of(seqOps, [](auto &&op) { return isa<EnableOp>(op); }) &&
+         "The SeqOp should only contain EnableOps in this pass.");
+  // This should be the number of enable statements + 1 since this is the
+  // maximum value the FSM register will reach.
+  size_t fsmBitWidth = getNecessaryBitWidth(seqOps.size() + 1);
+
+  OpBuilder builder(component->getRegion(0));
+  auto fsmRegister = createRegister(builder, component, fsmBitWidth, "fsm");
+  auto fsmIn = fsmRegister.getResult(0);
+  auto fsmWriteEn = fsmRegister.getResult(1);
+  auto fsmOut = fsmRegister.getResult(4);
+
+  builder.setInsertionPointToStart(wiresBody);
+  auto oneConstant = createConstant(builder, wires->getLoc(), 1, 1);
+
+  // Create the new compilation group to replace this SeqOp.
+  builder.setInsertionPointToEnd(wiresBody);
+  auto seqGroup =
+      builder.create<GroupOp>(wires->getLoc(), builder.getStringAttr("seq"));
+  Block *seqGroupBody = new Block();
+  seqGroup->getRegion(0).push_back(seqGroupBody);
+
+  // Guarantees a unique SymbolName for the group.
+  SymbolTable symTable(wires);
+  symTable.insert(seqGroup);
+
+  size_t fsmIndex = 0;
+  SmallVector<Attribute, 8> groupNames;
+  Value fsmNextState;
+  seq.walk([&](EnableOp enable) {
+    StringRef groupName = enable.groupName();
+    groupNames.push_back(SymbolRefAttr::get(builder.getContext(), groupName));
+    auto groupOp = cast<GroupOp>(symTable.lookup(groupName));
```

Comment:
I believe you can just do `wires.lookupSymbol(...)`? And potentially obviate the need for `symTable` at all? Also--there should be a templated method `wires.lookupSymbol<GroupOp>` that avoids the cast.